tests/evm/test_end_block.py

```diff
@@ -52,6 +59,40 @@ def test_end_block(is_last_step: bool, empty_block: bool, max_txs: int):
                     value=FQ(tx.id),
                 )
             )
+            # append receipt
```

Comment:
Ah, I now see why you added a `+1` to the rw_counter in the `CumulativeGasUsed` lookup.
If I understand it correctly, you're trying to reuse the rw entries from `end_tx` in the `end_block` right?
If that's the case, this won't always work because we may have an empty block which doesn't contain the `end_tx` state.
So the `end_block` step should have their own rw entries, and they should only be 2: `RWTableTag.CallContext` and `TxReceiptFieldTag.CumulativeGasUsed`.

Reply:
I also think only reading `TxReceiptFieldTag.CumulativeGasUsed` should work (so it adds its own RwTable entry). But we might need to additionally check whether the value is zero if it's an empty block, since there is no constraint on the `CumulativeGasUsed` value if it's the first read (the value seems not important now, but when someday we need to compute the block hash in circuit, we'd need to connect this `CumulativeGasUsed` value to block header field somewhere).

Reply:
> I think in the case of having an empty block it will still work because the check is only done for non-empty blocks. However I think your point is still valid. 

Ah you're right, if the block is empty, not even a single `TxReceiptFieldTag.CumulativeGasUsed` rw operation can be made because there's no previous tx!  But as you say, this is not a problem because the circuit only checks this lookup in the non-empty block case.

> Let me know if I got this right: `EndBlock` should add its own RwTable entry with `CumulativeGasUsed` but not the other 2 `TxReceipt` fields. This adds one extra row to the `RwTable` but removes the need to rely on `end_tx` state right? It also removes that magical `+1` I added earlier.

 correct!

> I also think only reading `TxReceiptFieldTag.CumulativeGasUsed` should work (so it adds its own RwTable entry). But we might need to additionally check whether the value is zero if it's an empty block, since there is no constraint on the `CumulativeGasUsed` value if it's the first read (the value seems not important now, but when someday we need to compute the block hash in circuit, we'd need to connect this `CumulativeGasUsed` value to block header field somewhere).

Ah, I think this needs to be resolved by:
1. Adding a new entry to the block table: `GasUsed`
2. In EndTx, if block is empty, lookup to block_table with `(GasUsed, 0)`.  If block is not empty, lookup to block_table with `(GasUsed, cumulative_gas)`

But we can address this in a future PR if you want :)

Reply:
> Ah, I think this needs to be resolved by:
> 1. Adding a new entry to the block table: `GasUsed`
> 2. In EndTx, if block is empty, lookup to block_table with `(GasUsed, 0)`.  If block is not empty, lookup to block_table with `(GasUsed, cumulative_gas)`
> 
> But we can address this in a future PR if you want :)

Make sense! I think we should not consider this for now and leave it in future PR, just want to remind that we need to check the value to be zero when it's empty block :)